crates/common/src/lib.rs

```diff
@@ -426,47 +426,49 @@ impl std::fmt::Display for Chain {
     }
 }
 
-#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize, Dummy)]
-pub struct StarknetVersion(String);
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Dummy)]
+pub struct StarknetVersion(u8, u8, u8, u8);
 
 impl StarknetVersion {
-    pub fn new(major: u64, minor: u64, patch: u64) -> Self {
-        StarknetVersion(format!("{major}.{minor}.{patch}"))
+    pub const fn new(a: u8, b: u8, c: u8, d: u8) -> Self {
+        StarknetVersion(a, b, c, d)
     }
+}
 
-    /// Parses the version string.
-    ///
-    /// Note: there are known deviations from semver such as version 0.11.0.2, which
-    /// will be truncated to 0.11.0 to still allow for parsing.
-    pub fn parse_as_semver(&self) -> anyhow::Result<Option<semver::Version>> {
-        // Truncate the 4th segment if present. This is a work-around for semver violating
-        // version strings like `0.11.0.2`.
-        let str = if self.0.is_empty() {
-            return Ok(None);
-        } else {
-            &self.0
-        };
-        let truncated = str
-            .match_indices('.')
-            .nth(2)
-            .map(|(index, _)| str.split_at(index).0)
-            .unwrap_or(str);
+impl FromStr for StarknetVersion {
+    type Err = anyhow::Error;
 
-        Some(semver::Version::parse(truncated).context("Parsing semver string")).transpose()
-    }
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Ok(StarknetVersion::new(0, 0, 0, 0));
+        }
 
-    pub fn as_str(&self) -> &str {
-        &self.0
-    }
+        let parts: Vec<_> = s.split('.').collect();
+        anyhow::ensure!(
+            parts.len() == 3 || parts.len() == 4,
+            "Invalid version string, expected 3 or 4 parts but got {}",
+            parts.len()
+        );
 
-    pub fn take_inner(self) -> String {
-        self.0
+        let a = parts[0].parse()?;
+        let b = parts[1].parse()?;
+        let c = parts[2].parse()?;
+        let d = parts.get(3).map(|x| x.parse()).transpose()?.unwrap_or(0);
+
+        Ok(StarknetVersion(a, b, c, d))
     }
 }
 
-impl From<String> for StarknetVersion {
-    fn from(value: String) -> Self {
-        Self(value)
+impl Display for StarknetVersion {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.0 == 0 && self.1 == 0 && self.2 == 0 && self.3 == 0 {
+            return Ok(());
+        }
```

Comment:
Ah nice, was wondering if you might leave this out. It kind of sucks that we have this "optional" state.